### ISSUE TITLE
Pod terminating stuck because of trying to umount not actual mounted dir

### ISSUE THREAD

**Issue by mochizuki875** [kubernetes/kubernetes#114546]
### What happened?

A status of Pod mounted NFS PV remain as terminating when a node shutdown and restart.

### What did you expect to happen?

A Pod is completely terminated.

### Summary

- Deploy a Pod mounted NFS PV
- Shutdown the node where the Pod is running
- Wait for 5 min and the Pod status become `Terminating`
- Restart the node
- The Pod status is still `Terminating` and the Pod is never terminated completely

<details>

### Reproduce

Prepare a NFS Server.(I used `nfs-kernel-server`)  

Deploy a Pod mounted NFS PV like this.  
`nfs-pv-test-pod.yaml`
```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    run: nfs-pv-test-pod
  name: nfs-pv-test-pod
spec:
  containers:
  - image: ubuntu
    name: nfs-pv-test-pod
    command: ["/bin/sh", "-c", "while :; do sleep 10; done"]
    volumeMounts:
      - name: nfs-test-volume
        mountPath: "/srv"
  volumes:
    - name: nfs-test-volume
      persistentVolumeClaim:
        claimName: nfs-test-pvc
  nodeSelector:
    kubernetes.io/hostname: node02

---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: nfs-test-pvc
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 1Mi

---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: nfs-test-pv
spec:
  capacity:
    storage: 1Mi
  accessModes:
    - ReadWriteMany
  nfs:
    server: <NFS Server IP Address>
    path: <NFS Server Dir Path>
```

Shutdown the node where the Pod is running.(In this case `node02`)
```console
root@node02:~# shutdown now
```

And the status of node02 becomes `NotReady`.
```console
$ k get no
NAME     STATUS     ROLES           AGE    VERSION
cp01     Ready      control-plane   126d   v1.25.0
node01   Ready      <none>          126d   v1.25.0
node02   NotReady   <none>          126d   v1.25.0
```

After 5 min later(by default), the Pod status becomes `Terminating`.  
(The time a Pod becomes `Terminating` status depends on `tolerationSeconds`.)  
https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#taint-based-evictions  
```console
$ k get po -owide -w 
NAME                    READY   STATUS        RESTARTS      AGE   IP            NODE     NOMINATED NODE   READINESS GATES
nfs-pv-test-pod         1/1     Running       0             27m   10.0.160.99   node02   <none>           <none>

nfs-pv-test-pod         1/1     Running       0             28m   10.0.160.99   node02   <none>           <none>
nfs-pv-test-pod         1/1     Terminating   0             33m   10.0.160.99   node02   <none>           <none>
nfs-pv-test-pod         0/1     Terminating   0             38m   <none>        node02   <none>           <none>
nfs-pv-test-pod         0/1     Terminating   0             39m   <none>        node02   <none>           <none>
```

Then, restart the node and the status of node02 becomes `Ready`.
```console
$ k get no
NAME     STATUS     ROLES           AGE    VERSION
cp01     Ready      control-plane   126d   v1.25.0
node01   Ready      <none>          126d   v1.25.0
node02   Ready      <none>          126d   v1.25.0
```

But the Pod status is still `Terminating` and the Pod is never terminated completely.
```console
$ k get po 
NAME                    READY   STATUS        RESTARTS      AGE   IP            NODE     NOMINATED NODE   READINESS GATES
nfs-pv-test-pod         0/1     Terminating   0             45m   <none>        node02   <none>           <none>
```

It seems that a `kubelet` faild to delete Pod because of trying to unmount a not mounted directory.
```console
$ k describe po nfs-pv-test-pod
Name:                      nfs-pv-test-pod
Namespace:                 default
...
Containers:
  nfs-pv-test-pod:
    Container ID:  containerd://c908fc752e70f5181b3983ce812e51be463cf3bd798d0a9b28cc45b3bb3d9353
    Image:         ubuntu
...
Events:
  Type     Reason         Age   From               Message
  ----     ------         ----  ----               -------
  Normal   Scheduled      10m   default-scheduler  Successfully assigned default/nfs-pv-test-pod to k8s-cluster3-worker02
  Normal   Pulling        10m   kubelet            Pulling image "ubuntu"
  Normal   Pulled         10m   kubelet            Successfully pulled image "ubuntu" in 6.353410672s
  Normal   Created        10m   kubelet            Created container nfs-pv-test-pod
  Normal   Started        10m   kubelet            Started container nfs-pv-test-pod
  Warning  NodeNotReady   9m5s  node-controller    Node is not ready
  Warning  FailedKillPod  56s   kubelet            error killing pod: failed to "KillPodSandbox" for "d695b232-ff16-454a-830b-02194bec6d01" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to destroy network for sandbox \"a078587c537484be9b622a2d1694cfaa3a6a2c6ac83872ca08c0c4758bb82af2\": plugin type=\"calico\" failed (delete): error getting ClusterInformation: Get \"https://10.96.0.1:443/apis/crd.projectcalico.org/v1/clusterinformations/default\": dial tcp 10.96.0.1:443: i/o timeout"
```

```console
root@node02:~# journalctl -xu kubelet.service
Dec 16 17:52:00 node02 kubelet[1151]: I1216 17:52:00.883395    1151 scope.go:115] "RemoveContainer" containerID="c908fc752e70f5181b3983ce812e51be463cf3bd798d0a9b28cc45b3bb3d9353"
...
Dec 16 17:52:40 node02 kubelet[1151]: I1216 17:52:40.628448    1151 reconciler.go:211] "operationExecutor.UnmountVolume started for volume \"nfs-test-pv\" (UniqueName: \"kubernetes.io/nfs/d695b232-ff16-454a-830b-02194bec6d01-nfs-test-pv\") pod \"d695b232-ff16-454a-830b-02194bec6d01\" (UID: \"d695b232-ff16-454a-830b-02194bec6d01\") "
Dec 16 17:52:40 node02 kubelet[1151]: E1216 17:52:40.629529    1151 nestedpendingoperations.go:335] Operation for "{volumeName:kubernetes.io/nfs/d695b232-ff16-454a-830b-02194bec6d01-nfs-test-pv podName:d695b232-ff16-454a-830b-02194bec6d01 nodeName:}" failed. No retries permitted until 2022-12-16 17:54:42.629513682 +0900 JST m=+252.044485589 (durationBeforeRetry 2m2s). Error: UnmountVolume.TearDown failed for volume "nfs-test-pv" (UniqueName: "kubernetes.io/nfs/d695b232-ff16-454a-830b-02194bec6d01-nfs-test-pv") pod "d695b232-ff16-454a-830b-02194bec6d01" (UID: "d695b232-ff16-454a-830b-02194bec6d01") : unmount failed: exit status 32
Dec 16 17:52:40 node02 kubelet[1151]: Unmounting arguments: /var/lib/kubelet/pods/d695b232-ff16-454a-830b-02194bec6d01/volumes/kubernetes.io~nfs/nfs-test-pv
Dec 16 17:52:40 node02 kubelet[1151]: Output: umount: /var/lib/kubelet/pods/d695b232-ff16-454a-830b-02194bec6d01/volumes/kubernetes.io~nfs/nfs-test-pv: not mounted.
```

### Delete Pod(Work Around)

Force delete Pod.
```console
$ kubectl delete po nfs-pv-test-pod --force --grace-period=0
Warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.
pod "nfs-pv-test-pod" force deleted
```

Delete Pod's directory on the node.
```console
root@node02:~# rm -rf /var/lib/kubelet/pods/d695b232-ff16-454a-830b-02194bec6d01
```

</details>

### Kubernetes version

<details>

```console
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:44:59Z", GoVersion:"go1.19", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:38:15Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
none(on premise)
</details>

### OS version

<details>

```console
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

$ uname -a
Linux k8s-cluster3-master01 5.15.0-53-generic #59-Ubuntu SMP Mon Oct 17 18:53:30 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>

```console
# kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:43:25Z", GoVersion:"go1.19", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Container runtime (CRI) and version (if applicable)

<details>

```console
# containerd --version
containerd containerd.io 1.6.8 9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6

# runc --version
runc version 1.1.4
commit: v1.1.4-0-g5fd4c4d
spec: 1.0.2-dev
go: go1.17.13
libseccomp: 2.5.3
```

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

```
none
```

</details>

**Comment by mochizuki875** [comment]
/sig storage
sig/ node

**Comment by mochizuki875** [comment]
It dose not happen using [kubernetes-csi/csi-driver-nfs](https://github.com/kubernetes-csi/csi-driver-nfs), so it seems to depend on NFS volume plugin.

**Comment by mochizuki875** [comment]
I try to build and run kubelet with comment out this logic.  
Then it seems to get well.

https://github.com/kubernetes/kubernetes/blob/0fdb14f63a83a830b8a7716da10fae5adcd2d927/staging/src/k8s.io/mount-utils/mount_helper_common.go#L55-L62

related:
#109676

This change checks the result of `umount` instead of the mount point check.
However, check logic is not implemented in `tryUnmount` and `forceUmount`(these functions are used in NFS volume plugin), despite it is done in `Unmount`.

https://github.com/kubernetes/kubernetes/blob/0fdb14f63a83a830b8a7716da10fae5adcd2d927/staging/src/k8s.io/mount-utils/mount_linux.go#L366-L378

https://github.com/kubernetes/kubernetes/blob/0fdb14f63a83a830b8a7716da10fae5adcd2d927/staging/src/k8s.io/mount-utils/mount_linux.go#L778-L806

**Comment by liggitt** [comment]
Reopening, fix had to be reverted due to unit tests requiring root permissions (https://github.com/kubernetes/kubernetes/pull/114605#discussion_r1104750857)

**Comment by mochizuki875** [comment]
@liggitt 

I am terribly sorry...
Please wait a moment.
I'll try to fix it.

**Comment by liggitt** [comment]
no worries, just get @jsafrane to review the rework, making sure the units tests don't require root permissions

**Comment by mochizuki875** [comment]
@liggitt Thank you! Fixed in #115769.

**Comment by saurav-agarwalla** [comment]
Are there any plans to backport this to older Kubernetes versions? This seems like a critical fix.
